### ISSUE TITLE
Can't publish ports from containers without EXPOSEd ports

### ISSUE THREAD

**Issue by davidgiesberg** [twitter-fabric/galley#23]
I'm puzzling through this right now and I'm not sure if this is a bug or a feature.

I feel like these should be equivalent:

```js
module.exports = {
  'alpha':{
    image: 'alpine',
    ports: ['9001:9001'],
    entrypoint: ['/bin/sleep'],
    command: ['100']
  }
}
```

```
docker run -it -p 9000:9000 alpine /bin/sleep 10
```

But, the one that I run through galley doesn't have the port published:
```
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                                         NAMES
c6c4204c2b46        alpine              "/bin/sleep 10"          4 seconds ago       Up 2 seconds        0.0.0.0:9000->9000/tcp                        amazing_fermat
2cb340eb6140        d6ead20d5571        "/bin/sleep 10"          5 seconds ago       Up 4 seconds                                                      alpha.dev
```

Am I missing something here?

If I build a container from the following Dockerfile, I can get galley to publish ports, whether or not they're the port exposed in the Dockerfile.

```
FROM alpine
EXPOSE 9001
```

I don't know node very well, but if I were to guess, it has something to do with this line:

https://github.com/twitter-fabric/galley/blob/master/lib/commands/run.coffee#L53

Any light you can shed would be great, thanks so much

**Comment by phopkins** [comment]
Yes, this looks like a bug. If you specify just `port: ['5000']` in the Galleyfile, then Galley will expose that port in all cases (letting it bind to a random port). However, the `port: [5000:5000]` only works if the port is `EXPOSE`d in the Dockerfile.

Does the workaround of `EXPOSE` in the Dockerfile suit you?

I'll put this on the list of things to fix, though also if you want to navigate the Docker API parameters (you found the right place in the code) I'm happy to look at a PR.

**Comment by davidgiesberg** [comment]
The EXPOSE workaround can work, but for my particular use case, I'm hoping
to expose different ports for different instances of the same image. I'll
give it a look this week and see what I can do with it.

**Comment by phopkins** [comment]
Cool. It may be that `PublishAllPorts` needs to be set more often.

Ideally the only ports that are exposed are those from the Galleyfile /
`galley` command-line, though if containers have some other ports exposed
that's not the worst, just as long as those ports end up on the
high-numbered, non-conflicting ports. (Given Galley's use in dev/test,
locked-down access isn't a huge concern, but port conflicts when running
multiple copies of the same image are.)

At any rate, thanks for trying out Galley! I hope it solves problems for
you.

**Comment by phopkins** [comment]
With the help of https://www.ctl.io/developers/blog/post/docker-networking-rules/ I sorted out expose and publish. Now fixed in 1.0.1.
